File: python/fetch-user.py
#coding: utf-8

import http.client
import zhihu
import dbhelper

conn = zhihu.get_conn()

# print("there are ",count(ids)," questions to fetch\n")

count = 0
insert_count = 0
while True:
    qid = dbhelper.next_question_id()
    if qid is None:
        break
    dbhelper.set_question_fetch(qid, dbhelper.FETCH_ING)
    url = "/question/{}".format(qid)
    conn.request("GET", url)
    response = conn.getresponse()
    if response is None:
        print('can not open', url)
    code = response.status
    print("{} [{}]".format(url, code))
    content = response.read()
    username_list = zhihu.get_username_list(content)

    for username, nickname in username_list.items():
        print("\t{:28s}{:8s}".format(username, nickname), end='')
        rs = dbhelper.saveUser(username, nickname)
        if rs is not None:
            insert_count += 1
            print('\t+')
        else:
            print()
        count += 1
    print("\tRate {}%\n".format(int(insert_count/count * 100)))
    dbhelper.set_question_fetch(qid, dbhelper.FETCH_OK)

conn.close()

File: fetch_user.php
<?php

defined('DEPLOY_MODE') or define('DEPLOY_MODE', 'DEV');

require_once ((__DIR__))."/vendor/autoload.php";
require_once __DIR__."/odie.php";
require_once __DIR__."/logic.php";
require_once __DIR__."/lib_mongodb.php";
require_once __DIR__."/autoload.php";

use model\User;
use model\Question;
use model\Answer;

DB::setAdapter(new adapter\Mysql());
// DB::setAdapter(new adapter\Mongo());

$base_url = 'http://www.zhihu.com';

$ids = Question::getIds();
echo "there are ",count($ids)," questions to fetch\n";

$sum = $existing_sum = 1;
foreach ($ids as $qid) {
    $url = "$base_url/question/$qid";
    echo $url,"\n";
    timer();
    list($code, $content) = odie_get($url);
    slog($url." [$code]");
    $t = timer();
    echo "Fetch question $qid [$code] $t ms\n";
    $username_list = get_username_list($content);

    timer();
    foreach ($username_list as $username => $nickname) {
        echo "\t$username";
        if (User::saveUser($username, $nickname)) {
            $existing_sum++;
        }
        $sum++;
    }
    Question::setFetched($qid);
    $rate = intval((1 - $existing_sum/$sum) * 100);
    $t = timer();
    echo "\tSave: $t ms\t Rate $rate%\n";
}


